src/common/components/HatsProtocol/ConnectFarcasterAccountViaHatsProtocol.tsx

```diff
@@ -0,0 +1,631 @@
+      },
+    };
+    const hash = hashTypedData(typedData);
+    const newSignature = await signTypedDataAsync(typedData);
+    const typeHash = keccak256(
+      toHex("SignedKeyRequest(uint256 requestFid,bytes key,uint256 deadline)")
+    );
+
+    const sig = encodePacked(
+      ["bytes", "bytes32", "uint256", "bytes", "uint256"],
+      [newSignature, typeHash, BigInt(fid), keccak256(key), newDeadline]
+    );
+
+    const result = await readContract(config, {
+      address: HatsFarcasterDelegatorContractAddress,
+      abi: HatsFarcasterDelegatorAbi,
+      functionName: "isValidSignature",
+      args: [hash, sig],
+    });
+
+    console.log("readContract result", result);
+    if (result === "0x1626ba7e") {
+      setSignature(newSignature);
+      setState(HatsProtocolSignupSteps[3]);
+    } else {
+      setState(HatsProtocolSignupSteps[5]);
+    }
+  };
+
+  const onAddHerocastSignerToHatsProtocol = async () => {
+    if (!address) return;
+
+    let hexStringPublicKey, hexStringPrivateKey;
+    if (
+      !hatsProtocolPendingAccounts ||
+      hatsProtocolPendingAccounts.length === 0
+    ) {
+      const { publicKey, privateKey } = await generateKeyPair();
+      hexStringPublicKey = bytesToHexString(publicKey)._unsafeUnwrap();
+      hexStringPrivateKey = bytesToHexString(privateKey)._unsafeUnwrap();
+
+      try {
+        addAccount({
+          id: null,
+          platformAccountId: fid.toString(),
+          status: AccountStatusType.pending,
+          platform: AccountPlatformType.farcaster_hats_protocol,
+          publicKey: hexStringPublicKey,
+          privateKey: hexStringPrivateKey,
+          data: {},
+        });
+      } catch (e) {
+        console.log("error when trying to add account", e);
+        setState(HatsProtocolSignupSteps[5]);
+        return;
+      }
+    } else {
+      hexStringPublicKey = hatsProtocolPendingAccounts[0].publicKey;
+      hexStringPrivateKey = hatsProtocolPendingAccounts[0].privateKey;
+    }
+
+    // const addressNounce = await readNoncesFromKeyGateway(address)
+    // console.log('KeyGateway: wallet nonce', addressNounce, 'delegatorContract nonce', nonce)
+    // const nonce = await readNoncesFromKeyGateway(address);
+    const nonce = await readNoncesFromKeyGateway(delegatorContractAddress);
+    const typedMetadataData = {
+      domain: SIGNED_KEY_REQUEST_VALIDATOR_EIP_712_DOMAIN,
+      types: {
+        SignedKeyRequest: SIGNED_KEY_REQUEST_TYPE,
+      },
+      primaryType: "SignedKeyRequest" as const,
+      message: {
+        requestFid: BigInt(fid),
+        key: hexStringPublicKey,
+        deadline: BigInt(deadline),
+      },
+    };
+    const metadataHash = hashTypedData(typedMetadataData);
+    const metadataSignature = await signTypedDataAsync(typedMetadataData);
+    console.log('metadataSignature', metadataSignature)
+
+    const METADATA_TYPEHASH = '0x16be47f1f1f50a66a48db64eba3fd35c21439c23622e513aab5b902018aec438';
+    const metadata = encodePacked(
```

Comment:
this might need to be `encode()` rather than `encodePacked()`

Reply:
more specifically, I think the last four values should be concatenated with `encode()`, and the result should be `encodePacked()` with the `metadataSignature`.